src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/CardinalityBasedGreedyJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,291 @@
+import se.liu.ida.hefquin.engine.federation.TPFServer;
+import se.liu.ida.hefquin.engine.federation.access.*;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.BRTPFRequestImpl;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.TPFRequestImpl;
+import se.liu.ida.hefquin.engine.federation.access.utils.FederationAccessUtils;
+import se.liu.ida.hefquin.engine.federation.access.utils.RequestMemberPair;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.*;
+import se.liu.ida.hefquin.engine.queryplan.utils.PhysicalPlanFactory;
+import se.liu.ida.hefquin.engine.queryproc.PhysicalOptimizationException;
+
+import java.util.*;
+
+/**
+ * This class implements a query optimizer that builds left-deep query plans,
+ * for which it uses a greedy approach to determine the join order based on cardinality estimation,
+ * and then choose physical algorithm according to the estimated number of request to execute the join.
+ */
+public class CardinalityBasedGreedyJoinPlanOptimizerImpl extends JoinPlanOptimizerBase
+{
+    protected final FederationAccessManager fedAccessMgr;
+
+    public CardinalityBasedGreedyJoinPlanOptimizerImpl(final FederationAccessManager fedAccessMgr ) {
+        this.fedAccessMgr = fedAccessMgr;
+    }
+
+    @Override
+	public EnumerationAlgorithm initializeEnumerationAlgorithm( final List<PhysicalPlan> subplans ) {
+		return new GreedyConstructionAlgorithm(subplans);
+	}
+
+	protected class GreedyConstructionAlgorithm implements EnumerationAlgorithm {
+        protected final List<PhysicalPlan> subplans;
+
+        public GreedyConstructionAlgorithm( final List<PhysicalPlan> subplans ) {
+            this.subplans = subplans;
+        }
+
+        public PhysicalPlan getResultingPlan() throws PhysicalOptimizationException {
+            final List<RequestMemberPair> flattenRequestMemPairs = new ArrayList<>();
+            for ( final PhysicalPlan plan : subplans ) {
+                flattenRequestMemPairs.addAll( createRequestMemPairsFromSourceAssignment(plan) );
+            }
+
+            final CardinalityResponse[] resps;
+            try {
+                resps = FederationAccessUtils.performCardinalityRequests(fedAccessMgr, flattenRequestMemPairs.toArray(new RequestMemberPair[0]));
+            } catch (final FederationAccessException e) {
+                throw new RuntimeException("Issuing a cardinality request caused an exception.", e);
+            }
+
+            final List<PhysicalPlanWithStatistics> subPlansWithStatistics = associateCardWithSubPlans( flattenRequestMemPairs, resps);
+
+            PhysicalPlanWithStatistics currentPlan = null;
+            for ( int k = 0; k < subplans.size(); k++ ) {
+                final PhysicalPlanWithStatistics nextPlan = chooseNextSubPlanBasedOnCard(subPlansWithStatistics);
```

Comment:
From this code, it is not immediately obvious that all that happens here is that the subplans are simply picked in order of their respective cardinality estimates. I would like that to be more obvious from the code itself.

You can simply sort the `subPlansWithStatistics` using the `sort` method of `Collections` with a `Comparator<PhysicalPlanWithStatistics>` that compares the `PhysicalPlanWithStatistics` objects based on their cardinality.

After that, you only need to iterate over that sorted list, which you can do with an `Iterator<PhysicalPlanWithStatistics>` where the first call of `next()` is used to initialize `currentPlan` and then you do `decidePhysicalAlgorithm` inside a while loop.

That would significantly simplify the code here and, also, make it easier to understand what is going on here. Additionally, that would avoid having to repeatedly go over all remaining elements of `subPlansWithStatistics` over and over again inside `chooseNextSubPlanBasedOnCard`, and it also avoids having to remove every `nextPlan` from `subPlansWithStatistics`.